quic/s2n-quic-transport/src/dc/manager.rs

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    contexts::WriteContext,
+    endpoint,
+    sync::{flag, flag::Writer},
+};
+use s2n_quic_core::{
+    ack,
+    crypto::tls,
+    dc,
+    dc::{Endpoint, Path},
+    ensure, event,
+    event::builder::{DcState, DcStateChanged},
+    frame::DcStatelessResetTokens,
+    packet::number::PacketNumber,
+    state::is,
+    stateless_reset, transmission,
+    transmission::interest::Query,
+};
+
+/// Manages transmission and receipt of `DC_STATELESS_RESET_TOKENS` and
+/// notifications to the dc::Endpoint Path
+pub struct Manager<Config: endpoint::Config> {
+    path: Option<<<Config as endpoint::Config>::DcEndpoint as Endpoint>::Path>,
+    version: Option<dc::Version>,
+    state: State,
+}
+
+type Flag = flag::Flag<DcStatelessResetTokenWriter>;
+
+#[derive(Debug, Default)]
+enum State {
+    #[default]
+    Init,
+    PathSecretsReady(Flag),
+    Complete,
+}
+
+impl State {
+    is!(is_init, Init);
+    is!(is_complete, Complete);
+    fn is_path_secrets_ready(&self) -> bool {
+        matches!(self, Self::PathSecretsReady(_))
+    }
+}
+
+impl<Config: endpoint::Config> Manager<Config> {
+    /// Constructs a new `dc::Manager` with the optional given path
+    ///
+    /// If path is `None`, the `dc::Manager` will be disabled
+    pub fn new<Pub: event::ConnectionPublisher>(
+        path: Option<<<Config as endpoint::Config>::DcEndpoint as Endpoint>::Path>,
+        version: dc::Version,
+        publisher: &mut Pub,
+    ) -> Self {
+        if path.is_some() {
+            publisher.on_dc_state_changed(DcStateChanged {
+                state: DcState::VersionNegotiated { version },
+            });
+            Self {
+                path,
+                version: Some(version),
+                state: State::Init,
+            }
+        } else {
+            Self::disabled()
+        }
+    }
+
+    /// Returns a disabled `dc::Manager`
+    pub fn disabled() -> Self {
+        Self {
+            path: None,
+            version: None,
+            state: State::Complete,
+        }
+    }
+
+    /// The dc version that was negotiated, if any
+    ///
+    /// Returns `None` if no version was negotiated or the `dc::Endpoint` did
+    /// not initialize a path for the connection
+    pub fn version(&self) -> Option<dc::Version> {
+        self.version
+    }
+
+    /// Called when the TLS session has indicated path secrets are ready
+    /// to be derived for the dc path
+    ///
+    /// Initiates sending of the `DC_STATELESS_RESET_TOKENS` frame
+    pub fn on_path_secrets_ready<Pub: event::ConnectionPublisher>(
+        &mut self,
+        session: &impl tls::TlsSession,
+        publisher: &mut Pub,
+    ) {
+        ensure!(!self.state.is_complete());
+
+        self.path.on_path_secrets_ready(session);
+        let mut flag = Flag::new(DcStatelessResetTokenWriter::new(
+            self.path.stateless_reset_tokens(),
```

Comment:
i think i mentioned in a previous PR that i think this method should return a `Vec<...>` rather than `&[...]`. Since we end up converting it to a Vec anyway here, we should do that.

Reply:
hmm, is there a benefit to doing that? I'd have to do some gating on `alloc` and I think it would just end up getting copied or cloned anyway?

Reply:
well my point is internally the provider would be creating its own `Vec` to be returned here. But since it would be borrowed, it'd need to set it on itself, only to be copied out here.

Reply:
ah got it, I wasn't thinking that the provider would be constructing the tokens when this method was called.

In that case, should I just have `on_path_secrets_ready` return the `Vec<stateless_reset::Token>`?

Reply:
oh that would work, yeah. the idea of the SRTs is that they need to be connected to the path secrets. so they definitely need to be created in response deriving the path secrets.